Sources/Runestone/TextView/TextInput/LayoutManager.swift

```diff
@@ -416,12 +416,20 @@ final class LayoutManager {
 
 // MARK: - UITextInput
 extension LayoutManager {
-    func caretRect(at location: Int) -> CGRect {
+    func caretRect(at location: Int, placeCaretAtNextLineFragmentForLastCharacter: Bool = false) -> CGRect {
         let safeLocation = min(max(location, 0), stringView.string.length)
         let line = lineManager.line(containingCharacterAt: safeLocation)!
         let lineController = lineControllerStorage.getOrCreateLineController(for: line)
-        let localLocation = safeLocation - line.location
-        let localCaretRect = lineController.caretRect(atIndex: localLocation)
+        let lineLocalLocation = safeLocation - line.location
+        if placeCaretAtNextLineFragmentForLastCharacter && lineLocalLocation > 0 && lineLocalLocation < line.data.totalLength {
+            // Special case: When we would otherwise place the caret at the very end of the line fragment, we move it to the beginning of the next line fragment. This special case is implemented to align with UITextView.
```

Comment:
This does something unexpected in my testing: If I'm using "Cmd + >" to navigate to the end of the line fragment, it jumps to the beginning of the next line (fragment), but I expect it to jump to the end of the line fragment. UITextView behaves as I expect.

However, this is made more tricky that the only way to place the caret at then of the line fragments in UIKit seems to be to use "Cmd + >", if I just navigate using arrow keys, it will never end up there, and will instead go to the next line. So the result of `caretRect` seems to not just depend on the selection itself but also _how_ that selection was made.

I'm typing this in Safari on macOS and it works the same way there, too. Good to test the behaviour.

(I'm sorry for opening this can of worms!)

Reply:
Thanks for testing this and reporting the issues. I've rolled back some of the changes and improved the the logic to differentiate between Ctrl+A/Ctrl+E and Cmd+Left/Cmd+Right. I believe this fixes the issues.

> (I'm sorry for opening this can of worms!)

Don't be! I've learned a lot about keyboard navigation, and in particular keyboard navigation with UITextInput, as I've been looking into this. I've also found and fixed one or two bugs along way 😊

Reply:
Good to hear, Simon.

This works as expected now in my testing and my muscle memory is happy. Very nice!